homeassistant/components/epson/media_player.py

```diff
@@ -82,22 +88,19 @@
 )
 
 
-async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
-    """Set up the Epson media player platform."""
-    if DATA_EPSON not in hass.data:
-        hass.data[DATA_EPSON] = []
-
-    name = config.get(CONF_NAME)
-    host = config.get(CONF_HOST)
-    port = config.get(CONF_PORT)
-    ssl = config[CONF_SSL]
-
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up the Epson projector from a config entry."""
+    timeout_scale = config_entry.options.get(TIMEOUT_SCALE, 1.0)
     epson_proj = EpsonProjector(
-        async_get_clientsession(hass, verify_ssl=False), name, host, port, ssl
+        async_get_clientsession(hass, verify_ssl=config_entry.data[CONF_SSL]),
+        config_entry.title,
+        config_entry.data[CONF_HOST],
+        config_entry.data[CONF_PORT],
+        timeout_scale,
+        config_entry.entry_id,
     )
-
-    hass.data[DATA_EPSON].append(epson_proj)
-    async_add_entities([epson_proj], update_before_add=True)
+    hass.data[DOMAIN][config_entry.entry_id][DATA_EPSON].append(epson_proj)
```

Comment:
You still use `hass.data[DOMAIN]` in `async_service_handler`. Looking at that function it seems not so straightforward to refactor to `hass.data[DOMAIN][config_entry.entry_id`

It also may be better to register service in `__init__.py` and I'm not sure if it's nice to access the entity from outside. Perhaps it's better though to refactor that in a separate PR not to make this too big.

Reply:
ok, I will create another PR later and move this service to `__init__.py`